{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy import spatial\n",
    "from sklearn.datasets import make_blobs\n",
    "from scipy.cluster.hierarchy import dendrogram, ward"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Purpose:  For the stromatolite project, identify which of the 10 flags:\n",
    "    1. Is most diverse (most number of unique ID's).\n",
    "    2. Is the most representitive (most similiar to others).\n",
    "    3. Is the most unique (most dissimiliar to others).\n",
    "    \n",
    "As X inputs use and steps to clean-up data:\n",
    "    1. Aggregated feature tables.\n",
    "        1.1. GroupBy 20s retention time window and 20 ppm m/z.\n",
    "        1.2. GroupBy flags.\n",
    "        1.3. X, Y, Z = m/z, t, area\n",
    "    2. GNPS ID's.\n",
    "        2.1 Summarize over flags.\n",
    "        2.2 Calculate pairwise similarity.\n",
    "        2.3 Ming: Library search for list of id's\n",
    "    3. 16S sequencing.\n",
    "        3.1 Summarize over flags.\n",
    "        3.2 Calculate pairwise similarity.\n",
    "        \n",
    "Sets = intersection, union, differences\n",
    "        \n",
    "Features: (Ming)\n",
    "    -MZmine:\n",
    "        -Aligns molecule 1 and molecule 2 (tolerance and ret time)\n",
    "        -Pick average or representitive sample instead of binning\n",
    "        -Assigns area and feature number\n",
    "        -Rows is feature versus column\n",
    "        -Could do post-clustering clean-up\n",
    "        -Most number of molecules least number of smaples:\n",
    "            -Set cover problem\n",
    "   \n",
    "   -Qiime:\n",
    "       -Filter to different taxonimic levels\n",
    "           X.X\n",
    "           X.X.X\n",
    "       -Easiest way to treat with set/cover:\n",
    "           First: Pick sample with most bacteria\n",
    "           Second: Pick sample with most unique bacteria\n",
    "           "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Part I: Genomic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = 'GBMF classification of OTUs f_Kerry McPhail stromatolite - GBMF.unique.good.pick.good.filt.tsv'\n",
    "dna16 = pd.read_csv(fn, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2375, 121)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna16.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxlevel</th>\n",
       "      <th>rankID</th>\n",
       "      <th>taxon</th>\n",
       "      <th>daughterlevels</th>\n",
       "      <th>total</th>\n",
       "      <th>GBMF1.groups</th>\n",
       "      <th>GBMF10.groups</th>\n",
       "      <th>GBMF100.groups</th>\n",
       "      <th>GBMF101.groups</th>\n",
       "      <th>GBMF102.groups</th>\n",
       "      <th>...</th>\n",
       "      <th>GBMF90.groups</th>\n",
       "      <th>GBMF91.groups</th>\n",
       "      <th>GBMF92.groups</th>\n",
       "      <th>GBMF93.groups</th>\n",
       "      <th>GBMF94.groups</th>\n",
       "      <th>GBMF95.groups</th>\n",
       "      <th>GBMF96.groups</th>\n",
       "      <th>GBMF97.groups</th>\n",
       "      <th>GBMF98.groups</th>\n",
       "      <th>GBMF99.groups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Root</td>\n",
       "      <td>2</td>\n",
       "      <td>12498</td>\n",
       "      <td>1455</td>\n",
       "      <td>1576</td>\n",
       "      <td>1177</td>\n",
       "      <td>966</td>\n",
       "      <td>829</td>\n",
       "      <td>...</td>\n",
       "      <td>1163</td>\n",
       "      <td>891</td>\n",
       "      <td>1325</td>\n",
       "      <td>1196</td>\n",
       "      <td>808</td>\n",
       "      <td>1088</td>\n",
       "      <td>1314</td>\n",
       "      <td>1420</td>\n",
       "      <td>1475</td>\n",
       "      <td>793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>50</td>\n",
       "      <td>12486</td>\n",
       "      <td>1454</td>\n",
       "      <td>1575</td>\n",
       "      <td>1177</td>\n",
       "      <td>966</td>\n",
       "      <td>829</td>\n",
       "      <td>...</td>\n",
       "      <td>1163</td>\n",
       "      <td>890</td>\n",
       "      <td>1325</td>\n",
       "      <td>1195</td>\n",
       "      <td>808</td>\n",
       "      <td>1088</td>\n",
       "      <td>1314</td>\n",
       "      <td>1420</td>\n",
       "      <td>1475</td>\n",
       "      <td>790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.1</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>20</td>\n",
       "      <td>272</td>\n",
       "      <td>77</td>\n",
       "      <td>98</td>\n",
       "      <td>28</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>41</td>\n",
       "      <td>12</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>27</td>\n",
       "      <td>35</td>\n",
       "      <td>68</td>\n",
       "      <td>62</td>\n",
       "      <td>53</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.10</td>\n",
       "      <td>Calditrichaeota</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxlevel  rankID            taxon  daughterlevels  total  GBMF1.groups  \\\n",
       "0         0       0             Root               2  12498          1455   \n",
       "1         1     0.1         Bacteria              50  12486          1454   \n",
       "2         1     0.2          unknown               1     12             1   \n",
       "3         2   0.1.1    Acidobacteria              20    272            77   \n",
       "4         2  0.1.10  Calditrichaeota               1     20             0   \n",
       "\n",
       "   GBMF10.groups  GBMF100.groups  GBMF101.groups  GBMF102.groups  ...  \\\n",
       "0           1576            1177             966             829  ...   \n",
       "1           1575            1177             966             829  ...   \n",
       "2              1               0               0               0  ...   \n",
       "3             98              28              20              13  ...   \n",
       "4              1               0               1               1  ...   \n",
       "\n",
       "   GBMF90.groups  GBMF91.groups  GBMF92.groups  GBMF93.groups  GBMF94.groups  \\\n",
       "0           1163            891           1325           1196            808   \n",
       "1           1163            890           1325           1195            808   \n",
       "2              0              1              0              1              0   \n",
       "3             41             12             48             48             27   \n",
       "4              0              0              0              0              0   \n",
       "\n",
       "   GBMF95.groups  GBMF96.groups  GBMF97.groups  GBMF98.groups  GBMF99.groups  \n",
       "0           1088           1314           1420           1475            793  \n",
       "1           1088           1314           1420           1475            790  \n",
       "2              0              0              0              0              3  \n",
       "3             35             68             62             53             16  \n",
       "4              0              0              1              0              0  \n",
       "\n",
       "[5 rows x 121 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna16.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['taxlevel', 'rankID', 'taxon', 'daughterlevels', 'total',\n",
       "       'GBMF1.groups', 'GBMF10.groups', 'GBMF100.groups', 'GBMF101.groups',\n",
       "       'GBMF102.groups',\n",
       "       ...\n",
       "       'GBMF90.groups', 'GBMF91.groups', 'GBMF92.groups', 'GBMF93.groups',\n",
       "       'GBMF94.groups', 'GBMF95.groups', 'GBMF96.groups', 'GBMF97.groups',\n",
       "       'GBMF98.groups', 'GBMF99.groups'],\n",
       "      dtype='object', length=121)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna16.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correlate columne name to flag: GBMFm.groups, where m = n, nn, or nnn\n",
    "# Run at different levels X.X ... X.X.X.X.X.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = 'StromTissueOnlyGNPS042119_quant - StromTissueOnlyGNPS042119_quant.tsv'\n",
    "raw_feat = pd.read_csv(fn, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row ID</th>\n",
       "      <th>row m/z</th>\n",
       "      <th>row retention time</th>\n",
       "      <th>correlation group ID</th>\n",
       "      <th>annotation network number</th>\n",
       "      <th>best ion</th>\n",
       "      <th>auto MS2 verify</th>\n",
       "      <th>identified by n=</th>\n",
       "      <th>partners</th>\n",
       "      <th>neutral M mass</th>\n",
       "      <th>...</th>\n",
       "      <th>Stromatolite_Tissue_99_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_91_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_79_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_95_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_82_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_80_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_1A_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_72_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_68_pos.mzXML Peak area</th>\n",
       "      <th>Stromatolite_Tissue_71_pos.mzXML Peak area</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>517.4830</td>\n",
       "      <td>5.8359</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>10500000.0</td>\n",
       "      <td>31700000.0</td>\n",
       "      <td>6.617186e+05</td>\n",
       "      <td>1.876589e+06</td>\n",
       "      <td>7.901106e+05</td>\n",
       "      <td>1.533882e+06</td>\n",
       "      <td>459000000.0</td>\n",
       "      <td>6.460571e+05</td>\n",
       "      <td>5.550346e+05</td>\n",
       "      <td>8.074183e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>539.4650</td>\n",
       "      <td>5.8358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>12300000.0</td>\n",
       "      <td>36200000.0</td>\n",
       "      <td>1.126247e+06</td>\n",
       "      <td>1.290824e+06</td>\n",
       "      <td>1.968257e+06</td>\n",
       "      <td>2.594918e+06</td>\n",
       "      <td>395000000.0</td>\n",
       "      <td>9.430503e+05</td>\n",
       "      <td>1.074441e+06</td>\n",
       "      <td>1.365752e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>871.5723</td>\n",
       "      <td>13.6490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>467000000.0</td>\n",
       "      <td>41600000.0</td>\n",
       "      <td>6.360000e+07</td>\n",
       "      <td>5.840498e+04</td>\n",
       "      <td>3.170000e+08</td>\n",
       "      <td>4.170000e+07</td>\n",
       "      <td>63800000.0</td>\n",
       "      <td>1.650000e+08</td>\n",
       "      <td>9.770000e+07</td>\n",
       "      <td>4.607939e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>537.3945</td>\n",
       "      <td>11.6977</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>357000000.0</td>\n",
       "      <td>306000000.0</td>\n",
       "      <td>3.530000e+08</td>\n",
       "      <td>4.770000e+08</td>\n",
       "      <td>3.150000e+08</td>\n",
       "      <td>2.870000e+08</td>\n",
       "      <td>241000000.0</td>\n",
       "      <td>2.110000e+08</td>\n",
       "      <td>2.460000e+08</td>\n",
       "      <td>3.270000e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>532.4393</td>\n",
       "      <td>11.6980</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>255000000.0</td>\n",
       "      <td>220000000.0</td>\n",
       "      <td>2.560000e+08</td>\n",
       "      <td>3.340000e+08</td>\n",
       "      <td>2.050000e+08</td>\n",
       "      <td>1.970000e+08</td>\n",
       "      <td>173000000.0</td>\n",
       "      <td>1.630000e+08</td>\n",
       "      <td>1.660000e+08</td>\n",
       "      <td>2.450000e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 140 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   row ID   row m/z  row retention time  correlation group ID  \\\n",
       "0       1  517.4830              5.8359                   NaN   \n",
       "1       2  539.4650              5.8358                   NaN   \n",
       "2       3  871.5723             13.6490                   NaN   \n",
       "3       4  537.3945             11.6977                   NaN   \n",
       "4       5  532.4393             11.6980                   NaN   \n",
       "\n",
       "   annotation network number  best ion  auto MS2 verify  identified by n=  \\\n",
       "0                        NaN       NaN              NaN               NaN   \n",
       "1                        NaN       NaN              NaN               NaN   \n",
       "2                        NaN       NaN              NaN               NaN   \n",
       "3                        NaN       NaN              NaN               NaN   \n",
       "4                        NaN       NaN              NaN               NaN   \n",
       "\n",
       "   partners  neutral M mass  ...  Stromatolite_Tissue_99_pos.mzXML Peak area  \\\n",
       "0       NaN             NaN  ...                                  10500000.0   \n",
       "1       NaN             NaN  ...                                  12300000.0   \n",
       "2       NaN             NaN  ...                                 467000000.0   \n",
       "3       NaN             NaN  ...                                 357000000.0   \n",
       "4       NaN             NaN  ...                                 255000000.0   \n",
       "\n",
       "   Stromatolite_Tissue_91_pos.mzXML Peak area  \\\n",
       "0                                  31700000.0   \n",
       "1                                  36200000.0   \n",
       "2                                  41600000.0   \n",
       "3                                 306000000.0   \n",
       "4                                 220000000.0   \n",
       "\n",
       "   Stromatolite_Tissue_79_pos.mzXML Peak area  \\\n",
       "0                                6.617186e+05   \n",
       "1                                1.126247e+06   \n",
       "2                                6.360000e+07   \n",
       "3                                3.530000e+08   \n",
       "4                                2.560000e+08   \n",
       "\n",
       "   Stromatolite_Tissue_95_pos.mzXML Peak area  \\\n",
       "0                                1.876589e+06   \n",
       "1                                1.290824e+06   \n",
       "2                                5.840498e+04   \n",
       "3                                4.770000e+08   \n",
       "4                                3.340000e+08   \n",
       "\n",
       "   Stromatolite_Tissue_82_pos.mzXML Peak area  \\\n",
       "0                                7.901106e+05   \n",
       "1                                1.968257e+06   \n",
       "2                                3.170000e+08   \n",
       "3                                3.150000e+08   \n",
       "4                                2.050000e+08   \n",
       "\n",
       "   Stromatolite_Tissue_80_pos.mzXML Peak area  \\\n",
       "0                                1.533882e+06   \n",
       "1                                2.594918e+06   \n",
       "2                                4.170000e+07   \n",
       "3                                2.870000e+08   \n",
       "4                                1.970000e+08   \n",
       "\n",
       "   Stromatolite_Tissue_1A_pos.mzXML Peak area  \\\n",
       "0                                 459000000.0   \n",
       "1                                 395000000.0   \n",
       "2                                  63800000.0   \n",
       "3                                 241000000.0   \n",
       "4                                 173000000.0   \n",
       "\n",
       "   Stromatolite_Tissue_72_pos.mzXML Peak area  \\\n",
       "0                                6.460571e+05   \n",
       "1                                9.430503e+05   \n",
       "2                                1.650000e+08   \n",
       "3                                2.110000e+08   \n",
       "4                                1.630000e+08   \n",
       "\n",
       "   Stromatolite_Tissue_68_pos.mzXML Peak area  \\\n",
       "0                                5.550346e+05   \n",
       "1                                1.074441e+06   \n",
       "2                                9.770000e+07   \n",
       "3                                2.460000e+08   \n",
       "4                                1.660000e+08   \n",
       "\n",
       "   Stromatolite_Tissue_71_pos.mzXML Peak area  \n",
       "0                                8.074183e+05  \n",
       "1                                1.365752e+06  \n",
       "2                                4.607939e+05  \n",
       "3                                3.270000e+08  \n",
       "4                                2.450000e+08  \n",
       "\n",
       "[5 rows x 140 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_feat.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['row ID', 'row m/z', 'row retention time', 'correlation group ID',\n",
       "       'annotation network number', 'best ion', 'auto MS2 verify',\n",
       "       'identified by n=', 'partners', 'neutral M mass',\n",
       "       'Stromatolite_Tissue_10_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_100_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_102_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_103_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_13_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_104_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_11_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_101_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_115_pos.mzXML Peak area',\n",
       "       'Stromatolite_Tissue_106_pos.mzXML Peak area'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_feat.columns[0:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correlate column name to flag: Stromatolite_Tissue_m_pos.mzXML, where m = n, nn, or nnn\n",
    "# Bin on retention time (20s) and m/z (20 ppm), areas to float."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = 'StromatoliteTissue_UpdateReDU-MS2V3.0_July19 - ReDUMS2_sample_information_template.tsv'\n",
    "meta = pd.read_csv(fn, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(154, 37)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unique ID's  job: RERUN Pool Region updated REDU v3 Schoenmakerskop Stromatolite Pool Samples 11/18 library search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correlate column name to flag: Stromatolite_Tissue_m_pos.mzXML, where m = n, nn, or nnn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = 'MOLECULAR-LIBRARYSEARCH-V2-402465f8-download_compound_occurrences-main.tsv'\n",
    "ids = pd.read_csv(fn, sep=\"\\t\")\n",
    "\n",
    "file_list = list(ids.columns)[2:]\n",
    "\n",
    "for f in file_list:\n",
    "    f1 = f.split('/')[-1]\n",
    "    f2 = f1.split('.mzML')[0]\n",
    "    ids.rename({f: f2}, axis=1, inplace=True)\n",
    "    \n",
    "head = list(ids.columns)[1:9]\n",
    "tail = list(ids.columns)[-11:-1]\n",
    "both = head + tail + ['blank_6']\n",
    "ids = ids.drop(columns=both)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = 'StromatoliteTissue_UpdateReDU-MS2V3.0_July19 - ReDUMS2_sample_information_template.tsv'\n",
    "meta = pd.read_csv(fn, sep=\"\\t\")\n",
    "meta['filename'] = meta['filename'].str.rstrip('.mzXML')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns = [1, 2, 3, 4, 5, 6, 7, 8, 10]\n",
    "for n in ns:\n",
    "    fs = list(meta[meta['Flag_Number'] == str(n)].filename)  \n",
    "    \n",
    "    # Present in redu file, but missing from Massive\n",
    "    missing_files = ['Stromatolite_Tissue_16_pos', \n",
    "                    'Stromatolite_Tissue_30_pos',\n",
    "                     'Stromatolite_Tissue_40_pos',\n",
    "                     'Stromatolite_Tissue_46_pos',\n",
    "                    'Stromatolite_Tissue_66_pos',\n",
    "                    'Stromatolite_Tissue_92_pos']   \n",
    "    for m in missing_files:\n",
    "        if m in fs:\n",
    "            fs.remove(m)\n",
    "      \n",
    "    label = 'site_' + str(n)\n",
    "    q = ids[fs]\n",
    "    ids[label] = ids[fs].sum(axis=1)\n",
    "    ids.drop(columns=fs, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def f(many, unique):\n",
    "    if many == 0 and unique > 0:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "# Total number of ID's\n",
    "number_ids = dict(zip(list(ids.columns)[1:], list(ids.sum())[1:]))\n",
    "\n",
    "# Total number of unique ID's\n",
    "ids.astype(bool).sum(axis=0)\n",
    "unique_ids = dict(zip(list(ids.columns)[1:], list(ids.astype(bool).sum(axis=0))[1:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'site_1': 19,\n",
       " 'site_2': 18,\n",
       " 'site_3': 63,\n",
       " 'site_4': 9,\n",
       " 'site_5': 15,\n",
       " 'site_6': 52,\n",
       " 'site_7': 24,\n",
       " 'site_8': 37,\n",
       " 'site_10': 36}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Leave one out, calculate # unique versus other sites...\n",
    "sites = ['site_1', 'site_2', 'site_3', 'site_4', 'site_5', 'site_6', 'site_7', 'site_8',\n",
    "         'site_10']\n",
    "sites2 = ['site_1', 'site_2', 'site_3', 'site_4', 'site_5', 'site_6', 'site_7', 'site_8',\n",
    "         'site_10']\n",
    "unique_vall = {}\n",
    "\n",
    "for s in sites:\n",
    "    ss = sites2\n",
    "    ss.remove(s)\n",
    "    ids['sums'] = ids[ss].sum(axis=1)\n",
    "    ids['temp'] = ids.apply(lambda x: f(x['sums'], x[s]), axis=1)\n",
    "    unique_vall[s] = ids.temp.sum()\n",
    "    ss = ss.append(s)\n",
    "    \n",
    "ids.drop(columns=['sums', 'temp'], inplace=True)\n",
    "unique_vall           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "bool_ids = ids.astype(bool).astype(int).copy(deep=True).drop(columns='LibraryID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_1</th>\n",
       "      <th>site_2</th>\n",
       "      <th>site_3</th>\n",
       "      <th>site_4</th>\n",
       "      <th>site_5</th>\n",
       "      <th>site_6</th>\n",
       "      <th>site_7</th>\n",
       "      <th>site_8</th>\n",
       "      <th>site_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.479</td>\n",
       "      <td>0.448</td>\n",
       "      <td>0.521</td>\n",
       "      <td>0.438</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.513</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0.479</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.568</td>\n",
       "      <td>0.614</td>\n",
       "      <td>0.625</td>\n",
       "      <td>0.561</td>\n",
       "      <td>0.615</td>\n",
       "      <td>0.585</td>\n",
       "      <td>0.523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0.448</td>\n",
       "      <td>0.568</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.553</td>\n",
       "      <td>0.545</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.610</td>\n",
       "      <td>0.640</td>\n",
       "      <td>0.640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0.521</td>\n",
       "      <td>0.614</td>\n",
       "      <td>0.553</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.555</td>\n",
       "      <td>0.579</td>\n",
       "      <td>0.628</td>\n",
       "      <td>0.542</td>\n",
       "      <td>0.557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0.438</td>\n",
       "      <td>0.625</td>\n",
       "      <td>0.545</td>\n",
       "      <td>0.555</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.532</td>\n",
       "      <td>0.603</td>\n",
       "      <td>0.605</td>\n",
       "      <td>0.551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.561</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.579</td>\n",
       "      <td>0.532</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.636</td>\n",
       "      <td>0.615</td>\n",
       "      <td>0.563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>0.513</td>\n",
       "      <td>0.615</td>\n",
       "      <td>0.610</td>\n",
       "      <td>0.628</td>\n",
       "      <td>0.603</td>\n",
       "      <td>0.636</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.653</td>\n",
       "      <td>0.603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.585</td>\n",
       "      <td>0.640</td>\n",
       "      <td>0.542</td>\n",
       "      <td>0.605</td>\n",
       "      <td>0.615</td>\n",
       "      <td>0.653</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.523</td>\n",
       "      <td>0.640</td>\n",
       "      <td>0.557</td>\n",
       "      <td>0.551</td>\n",
       "      <td>0.563</td>\n",
       "      <td>0.603</td>\n",
       "      <td>0.630</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site_1  site_2  site_3  site_4  site_5  site_6  site_7  site_8  site_10\n",
       "0   1.000   0.479   0.448   0.521   0.438   0.464   0.513   0.464    0.442\n",
       "1   0.479   1.000   0.568   0.614   0.625   0.561   0.615   0.585    0.523\n",
       "2   0.448   0.568   1.000   0.553   0.545   0.643   0.610   0.640    0.640\n",
       "3   0.521   0.614   0.553   1.000   0.555   0.579   0.628   0.542    0.557\n",
       "4   0.438   0.625   0.545   0.555   1.000   0.532   0.603   0.605    0.551\n",
       "5   0.464   0.561   0.643   0.579   0.532   1.000   0.636   0.615    0.563\n",
       "6   0.513   0.615   0.610   0.628   0.603   0.636   1.000   0.653    0.603\n",
       "7   0.464   0.585   0.640   0.542   0.605   0.615   0.653   1.000    0.630\n",
       "8   0.442   0.523   0.640   0.557   0.551   0.563   0.603   0.630    1.000"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# x = row, y = col [x, y]\n",
    "out_arr = np.zeros((9, 9))\n",
    "for x in range(0, 9):\n",
    "    for y in range(0, 9):\n",
    "        n = spatial.distance.cosine(bool_ids.iloc[:,x], bool_ids.iloc[:,y] )\n",
    "        out_arr[x, y] = 1 - n       \n",
    "\n",
    "cos = pd.DataFrame(out_arr).round(3)\n",
    "cos.columns = ['site_1', 'site_2', 'site_3', 'site_4', 'site_5',\n",
    "               'site_6', 'site_7', 'site_8', 'site_10']\n",
    "\n",
    "cos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sites</th>\n",
       "      <th>number_ids</th>\n",
       "      <th>unique_ids</th>\n",
       "      <th>unique_vall</th>\n",
       "      <th>cos_sum</th>\n",
       "      <th>cos_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>site_1</td>\n",
       "      <td>460</td>\n",
       "      <td>150</td>\n",
       "      <td>19</td>\n",
       "      <td>4.77</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>site_2</td>\n",
       "      <td>621</td>\n",
       "      <td>241</td>\n",
       "      <td>18</td>\n",
       "      <td>5.57</td>\n",
       "      <td>0.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>site_3</td>\n",
       "      <td>1632</td>\n",
       "      <td>417</td>\n",
       "      <td>63</td>\n",
       "      <td>5.65</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>site_4</td>\n",
       "      <td>524</td>\n",
       "      <td>186</td>\n",
       "      <td>9</td>\n",
       "      <td>5.55</td>\n",
       "      <td>0.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>site_5</td>\n",
       "      <td>725</td>\n",
       "      <td>239</td>\n",
       "      <td>15</td>\n",
       "      <td>5.45</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>site_6</td>\n",
       "      <td>934</td>\n",
       "      <td>342</td>\n",
       "      <td>52</td>\n",
       "      <td>5.59</td>\n",
       "      <td>0.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>site_7</td>\n",
       "      <td>936</td>\n",
       "      <td>295</td>\n",
       "      <td>24</td>\n",
       "      <td>5.86</td>\n",
       "      <td>0.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>site_8</td>\n",
       "      <td>1334</td>\n",
       "      <td>354</td>\n",
       "      <td>37</td>\n",
       "      <td>5.73</td>\n",
       "      <td>0.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>site_10</td>\n",
       "      <td>949</td>\n",
       "      <td>302</td>\n",
       "      <td>36</td>\n",
       "      <td>5.51</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sites  number_ids  unique_ids  unique_vall  cos_sum  cos_mean\n",
       "0   site_1         460         150           19     4.77      0.53\n",
       "1   site_2         621         241           18     5.57      0.62\n",
       "2   site_3        1632         417           63     5.65      0.63\n",
       "3   site_4         524         186            9     5.55      0.62\n",
       "4   site_5         725         239           15     5.45      0.61\n",
       "5   site_6         934         342           52     5.59      0.62\n",
       "6   site_7         936         295           24     5.86      0.65\n",
       "7   site_8        1334         354           37     5.73      0.64\n",
       "8  site_10         949         302           36     5.51      0.61"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cos_sum = dict(cos.sum())\n",
    "cos_mean = dict(cos.mean())\n",
    "\n",
    "out_df = pd.DataFrame()\n",
    "out_df['sites'] = number_ids.keys()\n",
    "out_df['number_ids'] = number_ids.values()\n",
    "out_df['unique_ids'] = unique_ids.values()\n",
    "out_df['unique_vall'] = unique_vall.values()\n",
    "out_df['cos_sum'] = cos_sum.values()\n",
    "out_df['cos_mean'] = cos_mean.values()\n",
    "out_df.set_index('sites')\n",
    "out_df.round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To do:  Analysis by features, and analysis by genomic stuff!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11384, 140)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_feat.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = list(raw_feat.columns)\n",
    "y = x[0:3]\n",
    "z = x[-130:-1]\n",
    "u = y + z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "proc_feat = raw_feat[u].copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = list(proc_feat.columns)[3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "replace_dict = {}\n",
    "for name in names:\n",
    "    n = name.split('.')[0]\n",
    "    replace_dict[name] = n  \n",
    "    \n",
    "proc_feat = proc_feat.rename(columns=replace_dict).copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "present_files = list(proc_feat.columns)\n",
    "final_feat = proc_feat.iloc[:,0:3].copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "flags = ['1', '2', '3', '4', '5', '6', '7', '8', '10']\n",
    "for f in flags:\n",
    "    col_to_pick = list(meta[meta.Flag_Number == f].filename)\n",
    "    \n",
    "    # Present in redu file, but missing from Massive\n",
    "    missing_files = ['Stromatolite_Tissue_16_pos', \n",
    "                    'Stromatolite_Tissue_30_pos',\n",
    "                     'Stromatolite_Tissue_40_pos',\n",
    "                     'Stromatolite_Tissue_46_pos',\n",
    "                    'Stromatolite_Tissue_66_pos',\n",
    "                    'Stromatolite_Tissue_92_pos']   \n",
    "    for m in missing_files:\n",
    "        if m in col_to_pick:\n",
    "            col_to_pick.remove(m)\n",
    "            \n",
    "    checked_col_to_pick = []\n",
    "    for c in col_to_pick:\n",
    "        if c in present_files:\n",
    "            checked_col_to_pick.append(c)\n",
    "    \n",
    "    out_title = 'sum_' + f\n",
    "\n",
    "    final_feat[out_title] = proc_feat[checked_col_to_pick].sum(axis=1, \n",
    "                                                               skipna=True,\n",
    "                                                              numeric_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row ID</th>\n",
       "      <th>row m/z</th>\n",
       "      <th>row retention time</th>\n",
       "      <th>sum_1</th>\n",
       "      <th>sum_2</th>\n",
       "      <th>sum_3</th>\n",
       "      <th>sum_4</th>\n",
       "      <th>sum_5</th>\n",
       "      <th>sum_6</th>\n",
       "      <th>sum_7</th>\n",
       "      <th>sum_8</th>\n",
       "      <th>sum_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>517.4830</td>\n",
       "      <td>5.8359</td>\n",
       "      <td>7.015500e+09</td>\n",
       "      <td>1.541656e+07</td>\n",
       "      <td>6.136707e+09</td>\n",
       "      <td>1.551042e+07</td>\n",
       "      <td>4.658652e+06</td>\n",
       "      <td>1.949041e+07</td>\n",
       "      <td>8.886632e+06</td>\n",
       "      <td>2.126596e+09</td>\n",
       "      <td>1.229202e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>539.4650</td>\n",
       "      <td>5.8358</td>\n",
       "      <td>5.185800e+09</td>\n",
       "      <td>1.762364e+07</td>\n",
       "      <td>4.486369e+09</td>\n",
       "      <td>1.974122e+07</td>\n",
       "      <td>4.931948e+06</td>\n",
       "      <td>2.186030e+07</td>\n",
       "      <td>1.345746e+07</td>\n",
       "      <td>2.109378e+09</td>\n",
       "      <td>1.347400e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>871.5723</td>\n",
       "      <td>13.6490</td>\n",
       "      <td>3.371560e+09</td>\n",
       "      <td>2.580225e+09</td>\n",
       "      <td>7.907935e+09</td>\n",
       "      <td>2.596622e+09</td>\n",
       "      <td>3.512200e+09</td>\n",
       "      <td>2.079400e+09</td>\n",
       "      <td>3.767300e+09</td>\n",
       "      <td>3.364949e+09</td>\n",
       "      <td>4.852518e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>537.3945</td>\n",
       "      <td>11.6977</td>\n",
       "      <td>3.978000e+09</td>\n",
       "      <td>3.715000e+09</td>\n",
       "      <td>8.377000e+09</td>\n",
       "      <td>3.310000e+09</td>\n",
       "      <td>4.006000e+09</td>\n",
       "      <td>2.980000e+09</td>\n",
       "      <td>3.852000e+09</td>\n",
       "      <td>5.886000e+09</td>\n",
       "      <td>4.576000e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>532.4393</td>\n",
       "      <td>11.6980</td>\n",
       "      <td>2.696000e+09</td>\n",
       "      <td>2.556000e+09</td>\n",
       "      <td>5.742000e+09</td>\n",
       "      <td>2.333000e+09</td>\n",
       "      <td>2.802000e+09</td>\n",
       "      <td>2.054000e+09</td>\n",
       "      <td>2.714000e+09</td>\n",
       "      <td>4.192000e+09</td>\n",
       "      <td>3.200000e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11379</td>\n",
       "      <td>38523</td>\n",
       "      <td>311.2220</td>\n",
       "      <td>4.5883</td>\n",
       "      <td>1.085796e+06</td>\n",
       "      <td>7.079029e+05</td>\n",
       "      <td>1.211532e+07</td>\n",
       "      <td>1.039420e+06</td>\n",
       "      <td>3.795534e+06</td>\n",
       "      <td>4.747803e+06</td>\n",
       "      <td>1.115058e+07</td>\n",
       "      <td>5.278053e+06</td>\n",
       "      <td>6.156807e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11380</td>\n",
       "      <td>38534</td>\n",
       "      <td>344.2801</td>\n",
       "      <td>3.9611</td>\n",
       "      <td>1.710442e+03</td>\n",
       "      <td>4.220484e+04</td>\n",
       "      <td>5.449430e+06</td>\n",
       "      <td>1.577832e+05</td>\n",
       "      <td>3.654567e+05</td>\n",
       "      <td>1.984119e+06</td>\n",
       "      <td>1.254403e+06</td>\n",
       "      <td>1.655166e+06</td>\n",
       "      <td>2.008885e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11381</td>\n",
       "      <td>38539</td>\n",
       "      <td>235.2058</td>\n",
       "      <td>4.5840</td>\n",
       "      <td>3.092209e+06</td>\n",
       "      <td>6.696389e+05</td>\n",
       "      <td>1.597207e+07</td>\n",
       "      <td>2.948499e+06</td>\n",
       "      <td>6.505277e+06</td>\n",
       "      <td>8.677586e+06</td>\n",
       "      <td>7.134641e+06</td>\n",
       "      <td>8.167086e+06</td>\n",
       "      <td>1.735757e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11382</td>\n",
       "      <td>38752</td>\n",
       "      <td>886.5476</td>\n",
       "      <td>5.9459</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.095750e+05</td>\n",
       "      <td>1.679539e+06</td>\n",
       "      <td>4.493016e+05</td>\n",
       "      <td>1.222971e+06</td>\n",
       "      <td>1.675469e+06</td>\n",
       "      <td>3.982267e+05</td>\n",
       "      <td>1.594338e+06</td>\n",
       "      <td>1.196960e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11383</td>\n",
       "      <td>39017</td>\n",
       "      <td>368.3169</td>\n",
       "      <td>6.6494</td>\n",
       "      <td>1.315027e+04</td>\n",
       "      <td>2.891399e+05</td>\n",
       "      <td>9.465673e+05</td>\n",
       "      <td>4.010595e+05</td>\n",
       "      <td>3.553592e+05</td>\n",
       "      <td>2.024039e+06</td>\n",
       "      <td>7.326719e+05</td>\n",
       "      <td>9.474029e+05</td>\n",
       "      <td>1.143632e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11384 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       row ID   row m/z  row retention time         sum_1         sum_2  \\\n",
       "0           1  517.4830              5.8359  7.015500e+09  1.541656e+07   \n",
       "1           2  539.4650              5.8358  5.185800e+09  1.762364e+07   \n",
       "2           3  871.5723             13.6490  3.371560e+09  2.580225e+09   \n",
       "3           4  537.3945             11.6977  3.978000e+09  3.715000e+09   \n",
       "4           5  532.4393             11.6980  2.696000e+09  2.556000e+09   \n",
       "...       ...       ...                 ...           ...           ...   \n",
       "11379   38523  311.2220              4.5883  1.085796e+06  7.079029e+05   \n",
       "11380   38534  344.2801              3.9611  1.710442e+03  4.220484e+04   \n",
       "11381   38539  235.2058              4.5840  3.092209e+06  6.696389e+05   \n",
       "11382   38752  886.5476              5.9459  0.000000e+00  5.095750e+05   \n",
       "11383   39017  368.3169              6.6494  1.315027e+04  2.891399e+05   \n",
       "\n",
       "              sum_3         sum_4         sum_5         sum_6         sum_7  \\\n",
       "0      6.136707e+09  1.551042e+07  4.658652e+06  1.949041e+07  8.886632e+06   \n",
       "1      4.486369e+09  1.974122e+07  4.931948e+06  2.186030e+07  1.345746e+07   \n",
       "2      7.907935e+09  2.596622e+09  3.512200e+09  2.079400e+09  3.767300e+09   \n",
       "3      8.377000e+09  3.310000e+09  4.006000e+09  2.980000e+09  3.852000e+09   \n",
       "4      5.742000e+09  2.333000e+09  2.802000e+09  2.054000e+09  2.714000e+09   \n",
       "...             ...           ...           ...           ...           ...   \n",
       "11379  1.211532e+07  1.039420e+06  3.795534e+06  4.747803e+06  1.115058e+07   \n",
       "11380  5.449430e+06  1.577832e+05  3.654567e+05  1.984119e+06  1.254403e+06   \n",
       "11381  1.597207e+07  2.948499e+06  6.505277e+06  8.677586e+06  7.134641e+06   \n",
       "11382  1.679539e+06  4.493016e+05  1.222971e+06  1.675469e+06  3.982267e+05   \n",
       "11383  9.465673e+05  4.010595e+05  3.553592e+05  2.024039e+06  7.326719e+05   \n",
       "\n",
       "              sum_8        sum_10  \n",
       "0      2.126596e+09  1.229202e+09  \n",
       "1      2.109378e+09  1.347400e+09  \n",
       "2      3.364949e+09  4.852518e+09  \n",
       "3      5.886000e+09  4.576000e+09  \n",
       "4      4.192000e+09  3.200000e+09  \n",
       "...             ...           ...  \n",
       "11379  5.278053e+06  6.156807e+06  \n",
       "11380  1.655166e+06  2.008885e+06  \n",
       "11381  8.167086e+06  1.735757e+07  \n",
       "11382  1.594338e+06  1.196960e+06  \n",
       "11383  9.474029e+05  1.143632e+06  \n",
       "\n",
       "[11384 rows x 12 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_feat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "row ID                11384\n",
       "row m/z               11384\n",
       "row retention time    11384\n",
       "sum_1                 10417\n",
       "sum_2                 10763\n",
       "sum_3                 11247\n",
       "sum_4                 10793\n",
       "sum_5                 10861\n",
       "sum_6                 10939\n",
       "sum_7                 10944\n",
       "sum_8                 11095\n",
       "sum_10                10995\n",
       "dtype: int64"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_feat.astype('bool').sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_feat = final_feat.iloc[:,3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To calculate:\n",
    "# Number of ID's\n",
    "# Unique ID's versus all\n",
    "# Sum intensity\n",
    "# Similarity matrix versus other samples, avg and median."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_ids = dict(calc_feat.astype(bool).sum(axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "area_sum = dict(calc_feat.sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = list(calc_feat.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['sum_1', 'sum_2', 'sum_3', 'sum_4', 'sum_5', 'sum_6', 'sum_7', 'sum_8', 'sum_10']\n"
     ]
    }
   ],
   "source": [
    "print(cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_dict = {}\n",
    "for c in cols:\n",
    "    # Select non-zero, feature present, in specified column.\n",
    "    working = calc_feat[calc_feat[c] > 0].copy(deep=True)\n",
    "    \n",
    "    # Calculate the sum of area in all other columns.\n",
    "    working['sum_others'] = working[cols].sum(axis=1) - working[c]\n",
    "    \n",
    "    not_unique = working['sum_others'].astype(bool).sum(axis=0)\n",
    "    unique = list(working.shape)[0] - not_unique\n",
    "    unique_dict[c] = unique\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sum_1': 1,\n",
       " 'sum_2': 1,\n",
       " 'sum_3': 5,\n",
       " 'sum_4': 0,\n",
       " 'sum_5': 1,\n",
       " 'sum_6': 3,\n",
       " 'sum_7': 1,\n",
       " 'sum_8': 4,\n",
       " 'sum_10': 2}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Similarity matrix, copy code from above?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row ID</th>\n",
       "      <th>row m/z</th>\n",
       "      <th>row retention time</th>\n",
       "      <th>sum_1</th>\n",
       "      <th>sum_2</th>\n",
       "      <th>sum_3</th>\n",
       "      <th>sum_4</th>\n",
       "      <th>sum_5</th>\n",
       "      <th>sum_6</th>\n",
       "      <th>sum_7</th>\n",
       "      <th>sum_8</th>\n",
       "      <th>sum_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>517.4830</td>\n",
       "      <td>5.8359</td>\n",
       "      <td>7.015500e+09</td>\n",
       "      <td>1.541656e+07</td>\n",
       "      <td>6.136707e+09</td>\n",
       "      <td>1.551042e+07</td>\n",
       "      <td>4.658652e+06</td>\n",
       "      <td>1.949041e+07</td>\n",
       "      <td>8.886632e+06</td>\n",
       "      <td>2.126596e+09</td>\n",
       "      <td>1.229202e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>539.4650</td>\n",
       "      <td>5.8358</td>\n",
       "      <td>5.185800e+09</td>\n",
       "      <td>1.762364e+07</td>\n",
       "      <td>4.486369e+09</td>\n",
       "      <td>1.974122e+07</td>\n",
       "      <td>4.931948e+06</td>\n",
       "      <td>2.186030e+07</td>\n",
       "      <td>1.345746e+07</td>\n",
       "      <td>2.109378e+09</td>\n",
       "      <td>1.347400e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>871.5723</td>\n",
       "      <td>13.6490</td>\n",
       "      <td>3.371560e+09</td>\n",
       "      <td>2.580225e+09</td>\n",
       "      <td>7.907935e+09</td>\n",
       "      <td>2.596622e+09</td>\n",
       "      <td>3.512200e+09</td>\n",
       "      <td>2.079400e+09</td>\n",
       "      <td>3.767300e+09</td>\n",
       "      <td>3.364949e+09</td>\n",
       "      <td>4.852518e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>537.3945</td>\n",
       "      <td>11.6977</td>\n",
       "      <td>3.978000e+09</td>\n",
       "      <td>3.715000e+09</td>\n",
       "      <td>8.377000e+09</td>\n",
       "      <td>3.310000e+09</td>\n",
       "      <td>4.006000e+09</td>\n",
       "      <td>2.980000e+09</td>\n",
       "      <td>3.852000e+09</td>\n",
       "      <td>5.886000e+09</td>\n",
       "      <td>4.576000e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>532.4393</td>\n",
       "      <td>11.6980</td>\n",
       "      <td>2.696000e+09</td>\n",
       "      <td>2.556000e+09</td>\n",
       "      <td>5.742000e+09</td>\n",
       "      <td>2.333000e+09</td>\n",
       "      <td>2.802000e+09</td>\n",
       "      <td>2.054000e+09</td>\n",
       "      <td>2.714000e+09</td>\n",
       "      <td>4.192000e+09</td>\n",
       "      <td>3.200000e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11379</td>\n",
       "      <td>38523</td>\n",
       "      <td>311.2220</td>\n",
       "      <td>4.5883</td>\n",
       "      <td>1.085796e+06</td>\n",
       "      <td>7.079029e+05</td>\n",
       "      <td>1.211532e+07</td>\n",
       "      <td>1.039420e+06</td>\n",
       "      <td>3.795534e+06</td>\n",
       "      <td>4.747803e+06</td>\n",
       "      <td>1.115058e+07</td>\n",
       "      <td>5.278053e+06</td>\n",
       "      <td>6.156807e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11380</td>\n",
       "      <td>38534</td>\n",
       "      <td>344.2801</td>\n",
       "      <td>3.9611</td>\n",
       "      <td>1.710442e+03</td>\n",
       "      <td>4.220484e+04</td>\n",
       "      <td>5.449430e+06</td>\n",
       "      <td>1.577832e+05</td>\n",
       "      <td>3.654567e+05</td>\n",
       "      <td>1.984119e+06</td>\n",
       "      <td>1.254403e+06</td>\n",
       "      <td>1.655166e+06</td>\n",
       "      <td>2.008885e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11381</td>\n",
       "      <td>38539</td>\n",
       "      <td>235.2058</td>\n",
       "      <td>4.5840</td>\n",
       "      <td>3.092209e+06</td>\n",
       "      <td>6.696389e+05</td>\n",
       "      <td>1.597207e+07</td>\n",
       "      <td>2.948499e+06</td>\n",
       "      <td>6.505277e+06</td>\n",
       "      <td>8.677586e+06</td>\n",
       "      <td>7.134641e+06</td>\n",
       "      <td>8.167086e+06</td>\n",
       "      <td>1.735757e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11382</td>\n",
       "      <td>38752</td>\n",
       "      <td>886.5476</td>\n",
       "      <td>5.9459</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.095750e+05</td>\n",
       "      <td>1.679539e+06</td>\n",
       "      <td>4.493016e+05</td>\n",
       "      <td>1.222971e+06</td>\n",
       "      <td>1.675469e+06</td>\n",
       "      <td>3.982267e+05</td>\n",
       "      <td>1.594338e+06</td>\n",
       "      <td>1.196960e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11383</td>\n",
       "      <td>39017</td>\n",
       "      <td>368.3169</td>\n",
       "      <td>6.6494</td>\n",
       "      <td>1.315027e+04</td>\n",
       "      <td>2.891399e+05</td>\n",
       "      <td>9.465673e+05</td>\n",
       "      <td>4.010595e+05</td>\n",
       "      <td>3.553592e+05</td>\n",
       "      <td>2.024039e+06</td>\n",
       "      <td>7.326719e+05</td>\n",
       "      <td>9.474029e+05</td>\n",
       "      <td>1.143632e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11384 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       row ID   row m/z  row retention time         sum_1         sum_2  \\\n",
       "0           1  517.4830              5.8359  7.015500e+09  1.541656e+07   \n",
       "1           2  539.4650              5.8358  5.185800e+09  1.762364e+07   \n",
       "2           3  871.5723             13.6490  3.371560e+09  2.580225e+09   \n",
       "3           4  537.3945             11.6977  3.978000e+09  3.715000e+09   \n",
       "4           5  532.4393             11.6980  2.696000e+09  2.556000e+09   \n",
       "...       ...       ...                 ...           ...           ...   \n",
       "11379   38523  311.2220              4.5883  1.085796e+06  7.079029e+05   \n",
       "11380   38534  344.2801              3.9611  1.710442e+03  4.220484e+04   \n",
       "11381   38539  235.2058              4.5840  3.092209e+06  6.696389e+05   \n",
       "11382   38752  886.5476              5.9459  0.000000e+00  5.095750e+05   \n",
       "11383   39017  368.3169              6.6494  1.315027e+04  2.891399e+05   \n",
       "\n",
       "              sum_3         sum_4         sum_5         sum_6         sum_7  \\\n",
       "0      6.136707e+09  1.551042e+07  4.658652e+06  1.949041e+07  8.886632e+06   \n",
       "1      4.486369e+09  1.974122e+07  4.931948e+06  2.186030e+07  1.345746e+07   \n",
       "2      7.907935e+09  2.596622e+09  3.512200e+09  2.079400e+09  3.767300e+09   \n",
       "3      8.377000e+09  3.310000e+09  4.006000e+09  2.980000e+09  3.852000e+09   \n",
       "4      5.742000e+09  2.333000e+09  2.802000e+09  2.054000e+09  2.714000e+09   \n",
       "...             ...           ...           ...           ...           ...   \n",
       "11379  1.211532e+07  1.039420e+06  3.795534e+06  4.747803e+06  1.115058e+07   \n",
       "11380  5.449430e+06  1.577832e+05  3.654567e+05  1.984119e+06  1.254403e+06   \n",
       "11381  1.597207e+07  2.948499e+06  6.505277e+06  8.677586e+06  7.134641e+06   \n",
       "11382  1.679539e+06  4.493016e+05  1.222971e+06  1.675469e+06  3.982267e+05   \n",
       "11383  9.465673e+05  4.010595e+05  3.553592e+05  2.024039e+06  7.326719e+05   \n",
       "\n",
       "              sum_8        sum_10  \n",
       "0      2.126596e+09  1.229202e+09  \n",
       "1      2.109378e+09  1.347400e+09  \n",
       "2      3.364949e+09  4.852518e+09  \n",
       "3      5.886000e+09  4.576000e+09  \n",
       "4      4.192000e+09  3.200000e+09  \n",
       "...             ...           ...  \n",
       "11379  5.278053e+06  6.156807e+06  \n",
       "11380  1.655166e+06  2.008885e+06  \n",
       "11381  8.167086e+06  1.735757e+07  \n",
       "11382  1.594338e+06  1.196960e+06  \n",
       "11383  9.474029e+05  1.143632e+06  \n",
       "\n",
       "[11384 rows x 12 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_feat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "proc_feat = final_feat.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sum_1</th>\n",
       "      <th>sum_2</th>\n",
       "      <th>sum_3</th>\n",
       "      <th>sum_4</th>\n",
       "      <th>sum_5</th>\n",
       "      <th>sum_6</th>\n",
       "      <th>sum_7</th>\n",
       "      <th>sum_8</th>\n",
       "      <th>sum_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>7015500000</td>\n",
       "      <td>15416562</td>\n",
       "      <td>6136707321</td>\n",
       "      <td>15510417</td>\n",
       "      <td>4658651</td>\n",
       "      <td>19490412</td>\n",
       "      <td>8886631</td>\n",
       "      <td>2126595679</td>\n",
       "      <td>1229201542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>5185800000</td>\n",
       "      <td>17623637</td>\n",
       "      <td>4486369268</td>\n",
       "      <td>19741217</td>\n",
       "      <td>4931948</td>\n",
       "      <td>21860302</td>\n",
       "      <td>13457457</td>\n",
       "      <td>2109378309</td>\n",
       "      <td>1347400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3371559652</td>\n",
       "      <td>2580224794</td>\n",
       "      <td>7907934605</td>\n",
       "      <td>2596622085</td>\n",
       "      <td>3512199959</td>\n",
       "      <td>2079400000</td>\n",
       "      <td>3767300000</td>\n",
       "      <td>3364949401</td>\n",
       "      <td>4852518292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3978000000</td>\n",
       "      <td>3715000000</td>\n",
       "      <td>8377000000</td>\n",
       "      <td>3310000000</td>\n",
       "      <td>4006000000</td>\n",
       "      <td>2980000000</td>\n",
       "      <td>3852000000</td>\n",
       "      <td>5886000000</td>\n",
       "      <td>4576000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2696000000</td>\n",
       "      <td>2556000000</td>\n",
       "      <td>5742000000</td>\n",
       "      <td>2333000000</td>\n",
       "      <td>2802000000</td>\n",
       "      <td>2054000000</td>\n",
       "      <td>2714000000</td>\n",
       "      <td>4192000000</td>\n",
       "      <td>3200000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11379</td>\n",
       "      <td>1085795</td>\n",
       "      <td>707902</td>\n",
       "      <td>12115316</td>\n",
       "      <td>1039420</td>\n",
       "      <td>3795534</td>\n",
       "      <td>4747802</td>\n",
       "      <td>11150577</td>\n",
       "      <td>5278052</td>\n",
       "      <td>6156806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11380</td>\n",
       "      <td>1710</td>\n",
       "      <td>42204</td>\n",
       "      <td>5449430</td>\n",
       "      <td>157783</td>\n",
       "      <td>365456</td>\n",
       "      <td>1984118</td>\n",
       "      <td>1254402</td>\n",
       "      <td>1655166</td>\n",
       "      <td>2008884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11381</td>\n",
       "      <td>3092208</td>\n",
       "      <td>669638</td>\n",
       "      <td>15972070</td>\n",
       "      <td>2948498</td>\n",
       "      <td>6505276</td>\n",
       "      <td>8677586</td>\n",
       "      <td>7134640</td>\n",
       "      <td>8167086</td>\n",
       "      <td>17357571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11382</td>\n",
       "      <td>0</td>\n",
       "      <td>509575</td>\n",
       "      <td>1679538</td>\n",
       "      <td>449301</td>\n",
       "      <td>1222971</td>\n",
       "      <td>1675469</td>\n",
       "      <td>398226</td>\n",
       "      <td>1594337</td>\n",
       "      <td>1196959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11383</td>\n",
       "      <td>13150</td>\n",
       "      <td>289139</td>\n",
       "      <td>946567</td>\n",
       "      <td>401059</td>\n",
       "      <td>355359</td>\n",
       "      <td>2024039</td>\n",
       "      <td>732671</td>\n",
       "      <td>947402</td>\n",
       "      <td>1143631</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11384 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sum_1       sum_2       sum_3       sum_4       sum_5       sum_6  \\\n",
       "0      7015500000    15416562  6136707321    15510417     4658651    19490412   \n",
       "1      5185800000    17623637  4486369268    19741217     4931948    21860302   \n",
       "2      3371559652  2580224794  7907934605  2596622085  3512199959  2079400000   \n",
       "3      3978000000  3715000000  8377000000  3310000000  4006000000  2980000000   \n",
       "4      2696000000  2556000000  5742000000  2333000000  2802000000  2054000000   \n",
       "...           ...         ...         ...         ...         ...         ...   \n",
       "11379     1085795      707902    12115316     1039420     3795534     4747802   \n",
       "11380        1710       42204     5449430      157783      365456     1984118   \n",
       "11381     3092208      669638    15972070     2948498     6505276     8677586   \n",
       "11382           0      509575     1679538      449301     1222971     1675469   \n",
       "11383       13150      289139      946567      401059      355359     2024039   \n",
       "\n",
       "            sum_7       sum_8      sum_10  \n",
       "0         8886631  2126595679  1229201542  \n",
       "1        13457457  2109378309  1347400000  \n",
       "2      3767300000  3364949401  4852518292  \n",
       "3      3852000000  5886000000  4576000000  \n",
       "4      2714000000  4192000000  3200000000  \n",
       "...           ...         ...         ...  \n",
       "11379    11150577     5278052     6156806  \n",
       "11380     1254402     1655166     2008884  \n",
       "11381     7134640     8167086    17357571  \n",
       "11382      398226     1594337     1196959  \n",
       "11383      732671      947402     1143631  \n",
       "\n",
       "[11384 rows x 9 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proc_feat.drop(columns=['row ID', 'row m/z', 'row retention time']).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_1</th>\n",
       "      <th>site_2</th>\n",
       "      <th>site_3</th>\n",
       "      <th>site_4</th>\n",
       "      <th>site_5</th>\n",
       "      <th>site_6</th>\n",
       "      <th>site_7</th>\n",
       "      <th>site_8</th>\n",
       "      <th>site_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.739</td>\n",
       "      <td>0.722</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0.739</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.951</td>\n",
       "      <td>0.121</td>\n",
       "      <td>0.174</td>\n",
       "      <td>0.190</td>\n",
       "      <td>0.138</td>\n",
       "      <td>0.159</td>\n",
       "      <td>0.170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0.722</td>\n",
       "      <td>0.951</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.131</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.138</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.121</td>\n",
       "      <td>0.131</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.724</td>\n",
       "      <td>0.339</td>\n",
       "      <td>0.428</td>\n",
       "      <td>0.323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.174</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.481</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.735</td>\n",
       "      <td>0.802</td>\n",
       "      <td>0.882</td>\n",
       "      <td>0.452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.190</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.724</td>\n",
       "      <td>0.735</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.558</td>\n",
       "      <td>0.691</td>\n",
       "      <td>0.593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.138</td>\n",
       "      <td>0.138</td>\n",
       "      <td>0.339</td>\n",
       "      <td>0.802</td>\n",
       "      <td>0.558</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.953</td>\n",
       "      <td>0.432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.159</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.428</td>\n",
       "      <td>0.882</td>\n",
       "      <td>0.691</td>\n",
       "      <td>0.953</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.170</td>\n",
       "      <td>0.172</td>\n",
       "      <td>0.323</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.593</td>\n",
       "      <td>0.432</td>\n",
       "      <td>0.511</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site_1  site_2  site_3  site_4  site_5  site_6  site_7  site_8  site_10\n",
       "0   1.000   0.739   0.722   0.031   0.051   0.071   0.046   0.059    0.093\n",
       "1   0.739   1.000   0.951   0.121   0.174   0.190   0.138   0.159    0.170\n",
       "2   0.722   0.951   1.000   0.131   0.179   0.203   0.138   0.165    0.172\n",
       "3   0.031   0.121   0.131   1.000   0.481   0.724   0.339   0.428    0.323\n",
       "4   0.051   0.174   0.179   0.481   1.000   0.735   0.802   0.882    0.452\n",
       "5   0.071   0.190   0.203   0.724   0.735   1.000   0.558   0.691    0.593\n",
       "6   0.046   0.138   0.138   0.339   0.802   0.558   1.000   0.953    0.432\n",
       "7   0.059   0.159   0.165   0.428   0.882   0.691   0.953   1.000    0.511\n",
       "8   0.093   0.170   0.172   0.323   0.452   0.593   0.432   0.511    1.000"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# x = row, y = col [x, y]\n",
    "out_arr = np.zeros((9, 9))\n",
    "for x in range(0, 9):\n",
    "    for y in range(0, 9):\n",
    "        n = spatial.distance.cosine(proc_feat.iloc[:,x], proc_feat.iloc[:,y] )\n",
    "        out_arr[x, y] = 1 - n       \n",
    "\n",
    "cos_feat = pd.DataFrame(out_arr).round(3)\n",
    "cos_feat.columns = ['site_1', 'site_2', 'site_3', 'site_4', 'site_5',\n",
    "               'site_6', 'site_7', 'site_8', 'site_10']\n",
    "\n",
    "cos_feat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sites</th>\n",
       "      <th>number_ids</th>\n",
       "      <th>area_sum</th>\n",
       "      <th>unique_dict</th>\n",
       "      <th>cos_sum</th>\n",
       "      <th>cos_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>site_1</td>\n",
       "      <td>460</td>\n",
       "      <td>1.630005e+11</td>\n",
       "      <td>19</td>\n",
       "      <td>2.81</td>\n",
       "      <td>0.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>site_2</td>\n",
       "      <td>621</td>\n",
       "      <td>2.288671e+11</td>\n",
       "      <td>18</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>site_3</td>\n",
       "      <td>1632</td>\n",
       "      <td>5.956921e+11</td>\n",
       "      <td>63</td>\n",
       "      <td>3.66</td>\n",
       "      <td>0.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>site_4</td>\n",
       "      <td>524</td>\n",
       "      <td>2.255433e+11</td>\n",
       "      <td>9</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>site_5</td>\n",
       "      <td>725</td>\n",
       "      <td>2.678990e+11</td>\n",
       "      <td>15</td>\n",
       "      <td>4.76</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>site_6</td>\n",
       "      <td>934</td>\n",
       "      <td>3.846793e+11</td>\n",
       "      <td>52</td>\n",
       "      <td>4.76</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>site_7</td>\n",
       "      <td>936</td>\n",
       "      <td>4.560875e+11</td>\n",
       "      <td>24</td>\n",
       "      <td>4.41</td>\n",
       "      <td>0.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>site_8</td>\n",
       "      <td>1334</td>\n",
       "      <td>5.713464e+11</td>\n",
       "      <td>37</td>\n",
       "      <td>4.85</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>site_10</td>\n",
       "      <td>949</td>\n",
       "      <td>3.936328e+11</td>\n",
       "      <td>36</td>\n",
       "      <td>3.75</td>\n",
       "      <td>0.42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sites  number_ids      area_sum  unique_dict  cos_sum  cos_mean\n",
       "0   site_1         460  1.630005e+11           19     2.81      0.31\n",
       "1   site_2         621  2.288671e+11           18     3.64      0.40\n",
       "2   site_3        1632  5.956921e+11           63     3.66      0.41\n",
       "3   site_4         524  2.255433e+11            9     3.58      0.40\n",
       "4   site_5         725  2.678990e+11           15     4.76      0.53\n",
       "5   site_6         934  3.846793e+11           52     4.76      0.53\n",
       "6   site_7         936  4.560875e+11           24     4.41      0.49\n",
       "7   site_8        1334  5.713464e+11           37     4.85      0.54\n",
       "8  site_10         949  3.936328e+11           36     3.75      0.42"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cos_feat_sum = dict(cos_feat.sum())\n",
    "cos_feat_mean = dict(cos_feat.mean())\n",
    "\n",
    "out_feat_df = pd.DataFrame()\n",
    "out_feat_df['sites'] = number_ids.keys()\n",
    "out_feat_df['number_ids'] = number_ids.values()\n",
    "out_feat_df['area_sum'] = area_sum.values()\n",
    "out_feat_df['unique_dict'] = unique_vall.values()\n",
    "out_feat_df['cos_sum'] = cos_feat_sum.values()\n",
    "out_feat_df['cos_mean'] = cos_feat_mean.values()\n",
    "out_feat_df.set_index('sites')\n",
    "out_feat_df.round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genetic differences..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxlevel</th>\n",
       "      <th>rankID</th>\n",
       "      <th>taxon</th>\n",
       "      <th>daughterlevels</th>\n",
       "      <th>total</th>\n",
       "      <th>GBMF1.groups</th>\n",
       "      <th>GBMF10.groups</th>\n",
       "      <th>GBMF100.groups</th>\n",
       "      <th>GBMF101.groups</th>\n",
       "      <th>GBMF102.groups</th>\n",
       "      <th>...</th>\n",
       "      <th>GBMF90.groups</th>\n",
       "      <th>GBMF91.groups</th>\n",
       "      <th>GBMF92.groups</th>\n",
       "      <th>GBMF93.groups</th>\n",
       "      <th>GBMF94.groups</th>\n",
       "      <th>GBMF95.groups</th>\n",
       "      <th>GBMF96.groups</th>\n",
       "      <th>GBMF97.groups</th>\n",
       "      <th>GBMF98.groups</th>\n",
       "      <th>GBMF99.groups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Root</td>\n",
       "      <td>2</td>\n",
       "      <td>12498</td>\n",
       "      <td>1455</td>\n",
       "      <td>1576</td>\n",
       "      <td>1177</td>\n",
       "      <td>966</td>\n",
       "      <td>829</td>\n",
       "      <td>...</td>\n",
       "      <td>1163</td>\n",
       "      <td>891</td>\n",
       "      <td>1325</td>\n",
       "      <td>1196</td>\n",
       "      <td>808</td>\n",
       "      <td>1088</td>\n",
       "      <td>1314</td>\n",
       "      <td>1420</td>\n",
       "      <td>1475</td>\n",
       "      <td>793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>50</td>\n",
       "      <td>12486</td>\n",
       "      <td>1454</td>\n",
       "      <td>1575</td>\n",
       "      <td>1177</td>\n",
       "      <td>966</td>\n",
       "      <td>829</td>\n",
       "      <td>...</td>\n",
       "      <td>1163</td>\n",
       "      <td>890</td>\n",
       "      <td>1325</td>\n",
       "      <td>1195</td>\n",
       "      <td>808</td>\n",
       "      <td>1088</td>\n",
       "      <td>1314</td>\n",
       "      <td>1420</td>\n",
       "      <td>1475</td>\n",
       "      <td>790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.1</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>20</td>\n",
       "      <td>272</td>\n",
       "      <td>77</td>\n",
       "      <td>98</td>\n",
       "      <td>28</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>41</td>\n",
       "      <td>12</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>27</td>\n",
       "      <td>35</td>\n",
       "      <td>68</td>\n",
       "      <td>62</td>\n",
       "      <td>53</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.10</td>\n",
       "      <td>Calditrichaeota</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxlevel  rankID            taxon  daughterlevels  total  GBMF1.groups  \\\n",
       "0         0       0             Root               2  12498          1455   \n",
       "1         1     0.1         Bacteria              50  12486          1454   \n",
       "2         1     0.2          unknown               1     12             1   \n",
       "3         2   0.1.1    Acidobacteria              20    272            77   \n",
       "4         2  0.1.10  Calditrichaeota               1     20             0   \n",
       "\n",
       "   GBMF10.groups  GBMF100.groups  GBMF101.groups  GBMF102.groups  ...  \\\n",
       "0           1576            1177             966             829  ...   \n",
       "1           1575            1177             966             829  ...   \n",
       "2              1               0               0               0  ...   \n",
       "3             98              28              20              13  ...   \n",
       "4              1               0               1               1  ...   \n",
       "\n",
       "   GBMF90.groups  GBMF91.groups  GBMF92.groups  GBMF93.groups  GBMF94.groups  \\\n",
       "0           1163            891           1325           1196            808   \n",
       "1           1163            890           1325           1195            808   \n",
       "2              0              1              0              1              0   \n",
       "3             41             12             48             48             27   \n",
       "4              0              0              0              0              0   \n",
       "\n",
       "   GBMF95.groups  GBMF96.groups  GBMF97.groups  GBMF98.groups  GBMF99.groups  \n",
       "0           1088           1314           1420           1475            793  \n",
       "1           1088           1314           1420           1475            790  \n",
       "2              0              0              0              0              3  \n",
       "3             35             68             62             53             16  \n",
       "4              0              0              1              0              0  \n",
       "\n",
       "[5 rows x 121 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna16.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show overall table at 1 and 2 (cyanos)\n",
    "# Cos similarity \n",
    "# Sum at taxon level 0-n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = list(dna16.columns)[5:]\n",
    "\n",
    "replace_dict = {}\n",
    "for name in names:\n",
    "    n = name.split('GBMF')[1].split('.')[0]\n",
    "    n = 'Stromatolite_Tissue_' + n + '_pos'\n",
    "    replace_dict[name] = n  \n",
    "    \n",
    "proc_dna16 = dna16.rename(columns=replace_dict).copy(deep=True)\n",
    "\n",
    "ns = [1, 2, 3, 4, 5, 6, 7, 8, 10]\n",
    "for n in ns:\n",
    "    fs = list(meta[meta['Flag_Number'] == str(n)].filename)  \n",
    "\n",
    "    # Present in redu file, but missing from Massive\n",
    "\n",
    "    missing_files = ['Stromatolite_Tissue_1D_pos', 'Stromatolite_Tissue_1C_pos', \n",
    "    'Stromatolite_Tissue_1A_pos', 'Stromatolite_Tissue_2B-R_pos', \n",
    "    'Stromatolite_Tissue_2B-L_pos', 'Stromatolite_Tissue_2C-L_pos', \n",
    "    'Stromatolite_Tissue_2E-R_pos', 'Stromatolite_Tissue_2D-R_pos', \n",
    "    'Stromatolite_Tissue_2E-L_pos', 'Stromatolite_Tissue_2D-L_pos', \n",
    "    'Stromatolite_Tissue_2A-R_pos', 'Stromatolite_Tissue_1B_pos', \n",
    "    'Stromatolite_Tissue_2C-R_pos', 'Stromatolite_Tissue_1E_pos', \n",
    "    'Stromatolite_Tissue_2A-L_pos']   \n",
    "    \n",
    "    for m in missing_files:\n",
    "        if m in fs:\n",
    "            fs.remove(m)\n",
    "\n",
    "    label = 'site_' + str(n)\n",
    "\n",
    "    q = proc_dna16[fs]\n",
    "    \n",
    "    proc_dna16[label] = q.sum(axis=1)\n",
    "    proc_dna16.drop(columns=fs, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxlevel</th>\n",
       "      <th>rankID</th>\n",
       "      <th>taxon</th>\n",
       "      <th>daughterlevels</th>\n",
       "      <th>total</th>\n",
       "      <th>site_1</th>\n",
       "      <th>site_2</th>\n",
       "      <th>site_3</th>\n",
       "      <th>site_4</th>\n",
       "      <th>site_5</th>\n",
       "      <th>site_6</th>\n",
       "      <th>site_7</th>\n",
       "      <th>site_8</th>\n",
       "      <th>site_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Root</td>\n",
       "      <td>2</td>\n",
       "      <td>12498</td>\n",
       "      <td>17792</td>\n",
       "      <td>20516</td>\n",
       "      <td>16788</td>\n",
       "      <td>19263</td>\n",
       "      <td>18830</td>\n",
       "      <td>19643</td>\n",
       "      <td>18451</td>\n",
       "      <td>19883</td>\n",
       "      <td>14539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>50</td>\n",
       "      <td>12486</td>\n",
       "      <td>17788</td>\n",
       "      <td>20502</td>\n",
       "      <td>16780</td>\n",
       "      <td>19252</td>\n",
       "      <td>18816</td>\n",
       "      <td>19640</td>\n",
       "      <td>18448</td>\n",
       "      <td>19880</td>\n",
       "      <td>14535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.1</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>20</td>\n",
       "      <td>272</td>\n",
       "      <td>1057</td>\n",
       "      <td>738</td>\n",
       "      <td>538</td>\n",
       "      <td>706</td>\n",
       "      <td>640</td>\n",
       "      <td>715</td>\n",
       "      <td>746</td>\n",
       "      <td>762</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.10</td>\n",
       "      <td>Calditrichaeota</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "      <td>26</td>\n",
       "      <td>9</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2370</td>\n",
       "      <td>6</td>\n",
       "      <td>0.1.7.6.2.1.4</td>\n",
       "      <td>uncultured</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>26</td>\n",
       "      <td>59</td>\n",
       "      <td>41</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2371</td>\n",
       "      <td>6</td>\n",
       "      <td>0.1.8.1.1.1.1</td>\n",
       "      <td>CK-2C2-2_ge</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2372</td>\n",
       "      <td>6</td>\n",
       "      <td>0.1.9.1.1.1.1</td>\n",
       "      <td>WCHB1-02_ge</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2373</td>\n",
       "      <td>6</td>\n",
       "      <td>0.2.1.1.1.1.1</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2374</td>\n",
       "      <td>6</td>\n",
       "      <td>0.2.1.1.1.2.1</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2375 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      taxlevel         rankID            taxon  daughterlevels  total  site_1  \\\n",
       "0            0              0             Root               2  12498   17792   \n",
       "1            1            0.1         Bacteria              50  12486   17788   \n",
       "2            1            0.2          unknown               1     12       4   \n",
       "3            2          0.1.1    Acidobacteria              20    272    1057   \n",
       "4            2         0.1.10  Calditrichaeota               1     20      11   \n",
       "...        ...            ...              ...             ...    ...     ...   \n",
       "2370         6  0.1.7.6.2.1.4       uncultured               0     21      19   \n",
       "2371         6  0.1.8.1.1.1.1      CK-2C2-2_ge               0      2       0   \n",
       "2372         6  0.1.9.1.1.1.1      WCHB1-02_ge               0      2       2   \n",
       "2373         6  0.2.1.1.1.1.1     unclassified               0      1       0   \n",
       "2374         6  0.2.1.1.1.2.1     unclassified               0     11       4   \n",
       "\n",
       "      site_2  site_3  site_4  site_5  site_6  site_7  site_8  site_10  \n",
       "0      20516   16788   19263   18830   19643   18451   19883    14539  \n",
       "1      20502   16780   19252   18816   19640   18448   19880    14535  \n",
       "2         14       8      11      14       3       3       3        4  \n",
       "3        738     538     706     640     715     746     762      433  \n",
       "4         27       6      28      26       9      31       1        6  \n",
       "...      ...     ...     ...     ...     ...     ...     ...      ...  \n",
       "2370       6      22      24      26      59      41      16        7  \n",
       "2371       4       3       2       3       1       0       2        5  \n",
       "2372       0       1       0       0       0       0       4        5  \n",
       "2373       0       0       0       1       0       0       0        1  \n",
       "2374      14       8      11      13       3       3       3        3  \n",
       "\n",
       "[2375 rows x 14 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proc_dna16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxlevel</th>\n",
       "      <th>rankID</th>\n",
       "      <th>taxon</th>\n",
       "      <th>daughterlevels</th>\n",
       "      <th>total</th>\n",
       "      <th>site_1</th>\n",
       "      <th>site_2</th>\n",
       "      <th>site_3</th>\n",
       "      <th>site_4</th>\n",
       "      <th>site_5</th>\n",
       "      <th>site_6</th>\n",
       "      <th>site_7</th>\n",
       "      <th>site_8</th>\n",
       "      <th>site_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.1</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>20</td>\n",
       "      <td>272</td>\n",
       "      <td>1057</td>\n",
       "      <td>738</td>\n",
       "      <td>538</td>\n",
       "      <td>706</td>\n",
       "      <td>640</td>\n",
       "      <td>715</td>\n",
       "      <td>746</td>\n",
       "      <td>762</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.10</td>\n",
       "      <td>Calditrichaeota</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "      <td>26</td>\n",
       "      <td>9</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.11</td>\n",
       "      <td>Chlamydiae</td>\n",
       "      <td>2</td>\n",
       "      <td>197</td>\n",
       "      <td>125</td>\n",
       "      <td>248</td>\n",
       "      <td>135</td>\n",
       "      <td>220</td>\n",
       "      <td>129</td>\n",
       "      <td>125</td>\n",
       "      <td>100</td>\n",
       "      <td>186</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.12</td>\n",
       "      <td>Chloroflexi</td>\n",
       "      <td>13</td>\n",
       "      <td>795</td>\n",
       "      <td>1339</td>\n",
       "      <td>1570</td>\n",
       "      <td>925</td>\n",
       "      <td>1415</td>\n",
       "      <td>1282</td>\n",
       "      <td>1155</td>\n",
       "      <td>1316</td>\n",
       "      <td>1099</td>\n",
       "      <td>906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.13</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>4</td>\n",
       "      <td>462</td>\n",
       "      <td>671</td>\n",
       "      <td>1184</td>\n",
       "      <td>1163</td>\n",
       "      <td>1029</td>\n",
       "      <td>1188</td>\n",
       "      <td>1055</td>\n",
       "      <td>1084</td>\n",
       "      <td>1263</td>\n",
       "      <td>933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.14</td>\n",
       "      <td>Dadabacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>24</td>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.15</td>\n",
       "      <td>Deinococcus-Thermus</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>31</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>36</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.16</td>\n",
       "      <td>Dependentiae</td>\n",
       "      <td>1</td>\n",
       "      <td>229</td>\n",
       "      <td>138</td>\n",
       "      <td>244</td>\n",
       "      <td>169</td>\n",
       "      <td>196</td>\n",
       "      <td>199</td>\n",
       "      <td>115</td>\n",
       "      <td>170</td>\n",
       "      <td>145</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.17</td>\n",
       "      <td>Elusimicrobia</td>\n",
       "      <td>6</td>\n",
       "      <td>83</td>\n",
       "      <td>184</td>\n",
       "      <td>146</td>\n",
       "      <td>71</td>\n",
       "      <td>82</td>\n",
       "      <td>93</td>\n",
       "      <td>90</td>\n",
       "      <td>78</td>\n",
       "      <td>97</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.18</td>\n",
       "      <td>Entotheonellaeota</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>31</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.19</td>\n",
       "      <td>Epsilonbacteraeota</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.2</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>6</td>\n",
       "      <td>195</td>\n",
       "      <td>661</td>\n",
       "      <td>394</td>\n",
       "      <td>283</td>\n",
       "      <td>383</td>\n",
       "      <td>352</td>\n",
       "      <td>527</td>\n",
       "      <td>502</td>\n",
       "      <td>469</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.20</td>\n",
       "      <td>FBP</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.21</td>\n",
       "      <td>FCPU426</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>33</td>\n",
       "      <td>31</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>23</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>24</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.22</td>\n",
       "      <td>Fibrobacteres</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "      <td>36</td>\n",
       "      <td>71</td>\n",
       "      <td>57</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>25</td>\n",
       "      <td>13</td>\n",
       "      <td>27</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.23</td>\n",
       "      <td>Firestonebacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.24</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>6</td>\n",
       "      <td>83</td>\n",
       "      <td>44</td>\n",
       "      <td>91</td>\n",
       "      <td>122</td>\n",
       "      <td>62</td>\n",
       "      <td>53</td>\n",
       "      <td>9</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.25</td>\n",
       "      <td>Fusobacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.26</td>\n",
       "      <td>GAL15</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.27</td>\n",
       "      <td>Gemmatimonadetes</td>\n",
       "      <td>8</td>\n",
       "      <td>72</td>\n",
       "      <td>267</td>\n",
       "      <td>239</td>\n",
       "      <td>166</td>\n",
       "      <td>236</td>\n",
       "      <td>227</td>\n",
       "      <td>240</td>\n",
       "      <td>241</td>\n",
       "      <td>196</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.28</td>\n",
       "      <td>Hydrogenedentes</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>61</td>\n",
       "      <td>119</td>\n",
       "      <td>92</td>\n",
       "      <td>104</td>\n",
       "      <td>106</td>\n",
       "      <td>101</td>\n",
       "      <td>98</td>\n",
       "      <td>101</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.29</td>\n",
       "      <td>Kiritimatiellaeota</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>48</td>\n",
       "      <td>94</td>\n",
       "      <td>83</td>\n",
       "      <td>106</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>91</td>\n",
       "      <td>90</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.3</td>\n",
       "      <td>AncK6</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.30</td>\n",
       "      <td>Latescibacteria</td>\n",
       "      <td>2</td>\n",
       "      <td>42</td>\n",
       "      <td>116</td>\n",
       "      <td>64</td>\n",
       "      <td>58</td>\n",
       "      <td>85</td>\n",
       "      <td>69</td>\n",
       "      <td>65</td>\n",
       "      <td>79</td>\n",
       "      <td>63</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.31</td>\n",
       "      <td>Lentisphaerae</td>\n",
       "      <td>3</td>\n",
       "      <td>56</td>\n",
       "      <td>34</td>\n",
       "      <td>31</td>\n",
       "      <td>64</td>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>37</td>\n",
       "      <td>20</td>\n",
       "      <td>35</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.32</td>\n",
       "      <td>Margulisbacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.33</td>\n",
       "      <td>Nitrospinae</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.34</td>\n",
       "      <td>Nitrospirae</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>98</td>\n",
       "      <td>59</td>\n",
       "      <td>51</td>\n",
       "      <td>61</td>\n",
       "      <td>78</td>\n",
       "      <td>89</td>\n",
       "      <td>84</td>\n",
       "      <td>66</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.35</td>\n",
       "      <td>Omnitrophicaeota</td>\n",
       "      <td>3</td>\n",
       "      <td>477</td>\n",
       "      <td>702</td>\n",
       "      <td>593</td>\n",
       "      <td>376</td>\n",
       "      <td>349</td>\n",
       "      <td>546</td>\n",
       "      <td>501</td>\n",
       "      <td>427</td>\n",
       "      <td>395</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.36</td>\n",
       "      <td>PAUC34f</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>23</td>\n",
       "      <td>24</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>24</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.37</td>\n",
       "      <td>Patescibacteria</td>\n",
       "      <td>9</td>\n",
       "      <td>357</td>\n",
       "      <td>206</td>\n",
       "      <td>284</td>\n",
       "      <td>159</td>\n",
       "      <td>227</td>\n",
       "      <td>212</td>\n",
       "      <td>320</td>\n",
       "      <td>177</td>\n",
       "      <td>218</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>34</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.38</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>9</td>\n",
       "      <td>879</td>\n",
       "      <td>1303</td>\n",
       "      <td>1377</td>\n",
       "      <td>1077</td>\n",
       "      <td>1432</td>\n",
       "      <td>1522</td>\n",
       "      <td>1391</td>\n",
       "      <td>1498</td>\n",
       "      <td>1371</td>\n",
       "      <td>930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.39</td>\n",
       "      <td>Poribacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.4</td>\n",
       "      <td>Armatimonadetes</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "      <td>103</td>\n",
       "      <td>65</td>\n",
       "      <td>73</td>\n",
       "      <td>52</td>\n",
       "      <td>49</td>\n",
       "      <td>43</td>\n",
       "      <td>38</td>\n",
       "      <td>42</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.40</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>4</td>\n",
       "      <td>3751</td>\n",
       "      <td>5005</td>\n",
       "      <td>6154</td>\n",
       "      <td>4721</td>\n",
       "      <td>6067</td>\n",
       "      <td>5520</td>\n",
       "      <td>5926</td>\n",
       "      <td>5398</td>\n",
       "      <td>5994</td>\n",
       "      <td>4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.41</td>\n",
       "      <td>Rokubacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>72</td>\n",
       "      <td>24</td>\n",
       "      <td>14</td>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>39</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.42</td>\n",
       "      <td>Schekmanbacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.43</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>7</td>\n",
       "      <td>201</td>\n",
       "      <td>76</td>\n",
       "      <td>284</td>\n",
       "      <td>416</td>\n",
       "      <td>253</td>\n",
       "      <td>220</td>\n",
       "      <td>137</td>\n",
       "      <td>111</td>\n",
       "      <td>236</td>\n",
       "      <td>466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.44</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.45</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>1</td>\n",
       "      <td>685</td>\n",
       "      <td>1395</td>\n",
       "      <td>1678</td>\n",
       "      <td>1521</td>\n",
       "      <td>1497</td>\n",
       "      <td>1513</td>\n",
       "      <td>1748</td>\n",
       "      <td>1590</td>\n",
       "      <td>1837</td>\n",
       "      <td>1339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>43</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.46</td>\n",
       "      <td>WPS-2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>44</td>\n",
       "      <td>23</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>31</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>44</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.47</td>\n",
       "      <td>WS1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.48</td>\n",
       "      <td>WS2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.49</td>\n",
       "      <td>WS4</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>20</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.5</td>\n",
       "      <td>BRC1</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>81</td>\n",
       "      <td>137</td>\n",
       "      <td>94</td>\n",
       "      <td>105</td>\n",
       "      <td>122</td>\n",
       "      <td>77</td>\n",
       "      <td>86</td>\n",
       "      <td>96</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>48</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.50</td>\n",
       "      <td>Zixibacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>63</td>\n",
       "      <td>40</td>\n",
       "      <td>10</td>\n",
       "      <td>38</td>\n",
       "      <td>35</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.6</td>\n",
       "      <td>Bacteria_unclassified</td>\n",
       "      <td>1</td>\n",
       "      <td>1117</td>\n",
       "      <td>728</td>\n",
       "      <td>895</td>\n",
       "      <td>806</td>\n",
       "      <td>877</td>\n",
       "      <td>860</td>\n",
       "      <td>861</td>\n",
       "      <td>746</td>\n",
       "      <td>696</td>\n",
       "      <td>665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.7</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>6</td>\n",
       "      <td>2019</td>\n",
       "      <td>2938</td>\n",
       "      <td>3444</td>\n",
       "      <td>3376</td>\n",
       "      <td>3386</td>\n",
       "      <td>3429</td>\n",
       "      <td>3967</td>\n",
       "      <td>3488</td>\n",
       "      <td>4162</td>\n",
       "      <td>2761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.8</td>\n",
       "      <td>CK-2C2-2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>52</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1.9</td>\n",
       "      <td>Caldiserica</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>53</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2.1</td>\n",
       "      <td>unknown_unclassified</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    taxlevel  rankID                  taxon  daughterlevels  total  site_1  \\\n",
       "3          2   0.1.1          Acidobacteria              20    272    1057   \n",
       "4          2  0.1.10        Calditrichaeota               1     20      11   \n",
       "5          2  0.1.11             Chlamydiae               2    197     125   \n",
       "6          2  0.1.12            Chloroflexi              13    795    1339   \n",
       "7          2  0.1.13          Cyanobacteria               4    462     671   \n",
       "8          2  0.1.14           Dadabacteria               1      6      17   \n",
       "9          2  0.1.15    Deinococcus-Thermus               1     10      31   \n",
       "10         2  0.1.16           Dependentiae               1    229     138   \n",
       "11         2  0.1.17          Elusimicrobia               6     83     184   \n",
       "12         2  0.1.18      Entotheonellaeota               1      5      31   \n",
       "13         2  0.1.19     Epsilonbacteraeota               1     16       2   \n",
       "14         2   0.1.2         Actinobacteria               6    195     661   \n",
       "15         2  0.1.20                    FBP               1      2       4   \n",
       "16         2  0.1.21                FCPU426               1     16      33   \n",
       "17         2  0.1.22          Fibrobacteres               2     54      36   \n",
       "18         2  0.1.23      Firestonebacteria               1      1       0   \n",
       "19         2  0.1.24             Firmicutes               6     83      44   \n",
       "20         2  0.1.25           Fusobacteria               1      4       0   \n",
       "21         2  0.1.26                  GAL15               1      2      11   \n",
       "22         2  0.1.27       Gemmatimonadetes               8     72     267   \n",
       "23         2  0.1.28        Hydrogenedentes               1     55      61   \n",
       "24         2  0.1.29     Kiritimatiellaeota               1     61      48   \n",
       "25         2   0.1.3                  AncK6               1      3      13   \n",
       "26         2  0.1.30        Latescibacteria               2     42     116   \n",
       "27         2  0.1.31          Lentisphaerae               3     56      34   \n",
       "28         2  0.1.32       Margulisbacteria               1     10       4   \n",
       "29         2  0.1.33            Nitrospinae               2      2       7   \n",
       "30         2  0.1.34            Nitrospirae               2     27      98   \n",
       "31         2  0.1.35       Omnitrophicaeota               3    477     702   \n",
       "32         2  0.1.36                PAUC34f               1     13      23   \n",
       "33         2  0.1.37        Patescibacteria               9    357     206   \n",
       "34         2  0.1.38         Planctomycetes               9    879    1303   \n",
       "35         2  0.1.39           Poribacteria               1      2       0   \n",
       "36         2   0.1.4        Armatimonadetes               5     42     103   \n",
       "37         2  0.1.40         Proteobacteria               4   3751    5005   \n",
       "38         2  0.1.41           Rokubacteria               1     13      72   \n",
       "39         2  0.1.42       Schekmanbacteria               1      1       0   \n",
       "40         2  0.1.43           Spirochaetes               7    201      76   \n",
       "41         2  0.1.44            Tenericutes               1     10       8   \n",
       "42         2  0.1.45        Verrucomicrobia               1    685    1395   \n",
       "43         2  0.1.46                  WPS-2               1      9      20   \n",
       "44         2  0.1.47                    WS1               1      1       0   \n",
       "45         2  0.1.48                    WS2               1      9      20   \n",
       "46         2  0.1.49                    WS4               1     13       0   \n",
       "47         2   0.1.5                   BRC1               1     74      81   \n",
       "48         2  0.1.50           Zixibacteria               1     29      63   \n",
       "49         2   0.1.6  Bacteria_unclassified               1   1117     728   \n",
       "50         2   0.1.7          Bacteroidetes               6   2019    2938   \n",
       "51         2   0.1.8               CK-2C2-2               1      2       0   \n",
       "52         2   0.1.9            Caldiserica               1      2       2   \n",
       "53         2   0.2.1   unknown_unclassified               1     12       4   \n",
       "\n",
       "    site_2  site_3  site_4  site_5  site_6  site_7  site_8  site_10  \n",
       "3      738     538     706     640     715     746     762      433  \n",
       "4       27       6      28      26       9      31       1        6  \n",
       "5      248     135     220     129     125     100     186      162  \n",
       "6     1570     925    1415    1282    1155    1316    1099      906  \n",
       "7     1184    1163    1029    1188    1055    1084    1263      933  \n",
       "8       12       8      12      10      21      24      18        5  \n",
       "9        2      13       5      11      25      11      36        9  \n",
       "10     244     169     196     199     115     170     145      163  \n",
       "11     146      71      82      93      90      78      97       69  \n",
       "12       2      10       8      10      17      16       8        1  \n",
       "13      12      14      22      16       1       4       9       12  \n",
       "14     394     283     383     352     527     502     469      215  \n",
       "15       4       2       2       0       2       1       9        3  \n",
       "16      31      23      14      23      11       8      24       26  \n",
       "17      71      57      35      35      25      13      27       81  \n",
       "18       0       1       0       0       0       0       0        0  \n",
       "19      91     122      62      53       9      33      49      155  \n",
       "20       2       1       2       0       2       1       3        0  \n",
       "21       3       3       3       6       6      12       2        1  \n",
       "22     239     166     236     227     240     241     196      113  \n",
       "23     119      92     104     106     101      98     101       84  \n",
       "24      94      83     106      83     100      91      90       91  \n",
       "25       8       4       5       5       5       6       2        1  \n",
       "26      64      58      85      69      65      79      63       73  \n",
       "27      31      64      25      28      37      20      35       82  \n",
       "28       3       6       3       1       5       1       5        3  \n",
       "29       8       3      10       8       3       2       0        0  \n",
       "30      59      51      61      78      89      84      66       36  \n",
       "31     593     376     349     546     501     427     395      249  \n",
       "32      24       6      18      24      19      15       9        8  \n",
       "33     284     159     227     212     320     177     218      122  \n",
       "34    1377    1077    1432    1522    1391    1498    1371      930  \n",
       "35       0       0       1       2       4       1       0        0  \n",
       "36      65      73      52      49      43      38      42       32  \n",
       "37    6154    4721    6067    5520    5926    5398    5994     4142  \n",
       "38      24      14      30      20      19      39       1        1  \n",
       "39       0       0       1       1       1       3       0        0  \n",
       "40     284     416     253     220     137     111     236      466  \n",
       "41       2       5       7       1      14       4       4        4  \n",
       "42    1678    1521    1497    1513    1748    1590    1837     1339  \n",
       "43      44      23      30      30      31      24      28       13  \n",
       "44       0       3       0       0       0       0       0        3  \n",
       "45      15       5      10      15      16       9       6        4  \n",
       "46      32      20      11      14       0       0       4       19  \n",
       "47     137      94     105     122      77      86      96       88  \n",
       "48      40      10      38      35       9      22      10       16  \n",
       "49     895     806     877     860     861     746     696      665  \n",
       "50    3444    3376    3386    3429    3967    3488    4162     2761  \n",
       "51       4       3       2       3       1       0       2        5  \n",
       "52       0       1       0       0       0       0       4        5  \n",
       "53      14       8      11      14       3       3       3        4  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proc_dna16[proc_dna16.taxlevel==2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "cos_dna = proc_dna16.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "cos_dna = cos_dna.drop(columns=['taxlevel', 'rankID', 'taxon', 'daughterlevels', 'total']).astype(int).copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_1</th>\n",
       "      <th>site_2</th>\n",
       "      <th>site_3</th>\n",
       "      <th>site_4</th>\n",
       "      <th>site_5</th>\n",
       "      <th>site_6</th>\n",
       "      <th>site_7</th>\n",
       "      <th>site_8</th>\n",
       "      <th>site_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0.998</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.998</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.998</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0.997</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.998</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.998</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site_1  site_2  site_3  site_4  site_5  site_6  site_7  site_8  site_10\n",
       "0   1.000   0.998   0.997   0.997   0.997   0.997   0.997   0.997    0.996\n",
       "1   0.998   1.000   0.998   1.000   0.999   0.998   0.999   0.998    0.999\n",
       "2   0.997   0.998   1.000   0.998   0.999   0.999   0.999   0.999    0.999\n",
       "3   0.997   1.000   0.998   1.000   1.000   0.999   0.999   0.998    0.998\n",
       "4   0.997   0.999   0.999   1.000   1.000   0.999   1.000   0.998    0.998\n",
       "5   0.997   0.998   0.999   0.999   0.999   1.000   0.999   0.999    0.997\n",
       "6   0.997   0.999   0.999   0.999   1.000   0.999   1.000   0.999    0.998\n",
       "7   0.997   0.998   0.999   0.998   0.998   0.999   0.999   1.000    0.998\n",
       "8   0.996   0.999   0.999   0.998   0.998   0.997   0.998   0.998    1.000"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# x = row, y = col [x, y]\n",
    "out_arr = np.zeros((9, 9))\n",
    "for x in range(0, 9):\n",
    "    for y in range(0, 9):\n",
    "        n = spatial.distance.cosine(cos_dna.iloc[:,x], cos_dna.iloc[:,y] )\n",
    "        out_arr[x, y] = 1 - n       \n",
    "\n",
    "cos_dna = pd.DataFrame(out_arr).round(3)\n",
    "cos_dna.columns = ['site_1', 'site_2', 'site_3', 'site_4', 'site_5',\n",
    "               'site_6', 'site_7', 'site_8', 'site_10']\n",
    "\n",
    "cos_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "un_dna = proc_dna16.copy(deep=True)\n",
    "un_dna = un_dna.drop(columns=['taxlevel', 'rankID', 'taxon', 'daughterlevels', 'total']).astype(int).copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_1</th>\n",
       "      <th>site_2</th>\n",
       "      <th>site_3</th>\n",
       "      <th>site_4</th>\n",
       "      <th>site_5</th>\n",
       "      <th>site_6</th>\n",
       "      <th>site_7</th>\n",
       "      <th>site_8</th>\n",
       "      <th>site_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>17792</td>\n",
       "      <td>20516</td>\n",
       "      <td>16788</td>\n",
       "      <td>19263</td>\n",
       "      <td>18830</td>\n",
       "      <td>19643</td>\n",
       "      <td>18451</td>\n",
       "      <td>19883</td>\n",
       "      <td>14539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>17788</td>\n",
       "      <td>20502</td>\n",
       "      <td>16780</td>\n",
       "      <td>19252</td>\n",
       "      <td>18816</td>\n",
       "      <td>19640</td>\n",
       "      <td>18448</td>\n",
       "      <td>19880</td>\n",
       "      <td>14535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1057</td>\n",
       "      <td>738</td>\n",
       "      <td>538</td>\n",
       "      <td>706</td>\n",
       "      <td>640</td>\n",
       "      <td>715</td>\n",
       "      <td>746</td>\n",
       "      <td>762</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "      <td>26</td>\n",
       "      <td>9</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2370</td>\n",
       "      <td>19</td>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>26</td>\n",
       "      <td>59</td>\n",
       "      <td>41</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2371</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2372</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2373</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2374</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2375 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      site_1  site_2  site_3  site_4  site_5  site_6  site_7  site_8  site_10\n",
       "0      17792   20516   16788   19263   18830   19643   18451   19883    14539\n",
       "1      17788   20502   16780   19252   18816   19640   18448   19880    14535\n",
       "2          4      14       8      11      14       3       3       3        4\n",
       "3       1057     738     538     706     640     715     746     762      433\n",
       "4         11      27       6      28      26       9      31       1        6\n",
       "...      ...     ...     ...     ...     ...     ...     ...     ...      ...\n",
       "2370      19       6      22      24      26      59      41      16        7\n",
       "2371       0       4       3       2       3       1       0       2        5\n",
       "2372       2       0       1       0       0       0       0       4        5\n",
       "2373       0       0       0       0       1       0       0       0        1\n",
       "2374       4      14       8      11      13       3       3       3        3\n",
       "\n",
       "[2375 rows x 9 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "un_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_dict = {}\n",
    "cols = un_dna.columns\n",
    "for c in cols:\n",
    "    # Select non-zero, feature present, in specified column.\n",
    "    working = un_dna[un_dna[c] > 0].copy(deep=True)\n",
    "    \n",
    "    # Calculate the sum of area in all other columns.\n",
    "    working['sum_others'] = working[cols].sum(axis=1) - working[c]\n",
    "    \n",
    "    not_unique = working['sum_others'].astype(bool).sum(axis=0)\n",
    "    unique = list(working.shape)[0] - not_unique\n",
    "    unique_dict[c] = unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'site_1': 48,\n",
       " 'site_2': 29,\n",
       " 'site_3': 17,\n",
       " 'site_4': 38,\n",
       " 'site_5': 9,\n",
       " 'site_6': 40,\n",
       " 'site_7': 10,\n",
       " 'site_8': 3,\n",
       " 'site_10': 14}"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
